pretext/codechat.py

```diff
@@ -0,0 +1,110 @@
+    html_files = set(
+        pathlib.Path(html_file).stem for html_file in glob.glob(dest_dir + "/*.html")
+    )
+
+    # lxml turns ``xml:id`` into the string below.
+    xml_ns = "{http://www.w3.org/XML/1998/namespace}"
+    xml_base_attrib = f"{xml_ns}base"
+    xml_id_attrib = f"{xml_ns}id"
+
+    # Define a loader which sets the ``xml:base`` of an xincluded element. While lxml `evidently used to do this in 2013 <https://stackoverflow.com/a/18158472/16038919>`_, a change eliminated this ability per some `dicussion <https://mail.gnome.org/archives/xml/2014-April/msg00015.html>`_, which included a rejected patch fixing this problem. `Current source <https://github.com/GNOME/libxml2/blob/master/xinclude.c#L1689>`_ lacks this patch.
+    def my_loader(href, parse, encoding=None, parser=None):
+        ret = lxml.ElementInclude._lxml_default_loader(href, parse, encoding, parser)
+        # The return value may not be an element.
+        if isinstance(ret, ET._Element):
+            ret.attrib[xml_base_attrib] = href
+        return ret
+
+    # Load the XML, performing xincludes using this loader.
+    huge_parser = ET.XMLParser(huge_tree=True)
+    src_tree = ET.parse(xml, parser=huge_parser)
+    lxml.ElementInclude.include(src_tree, loader=my_loader)
+
+    # Walk though every element with an xml ID.
+    for elem in src_tree.iterfind(f"//*[@{xml_id_attrib}]"):
+        # Consider only elemets whose ID produced an HTML file. TODO: use a walrus operator after Python 3.7 is EOL.
+        xml_id = elem.get(xml_id_attrib)
+        if xml_id in html_files:
+            # Store this discovered mapping between ID and output file.
+            #
+            # The `elem.base <https://lxml.de/api/lxml.etree._Element-class.html#base>`_ gives the URL of this file (which is correct due to the custom loader). Extract the path.
+            up = urllib.parse.urlparse(elem.base)
+            # If this isn't a ``file`` scheme (or an unspecified schema, which seems to default to a file), we're lost.
+            assert up.scheme in ("file", "")
+            path = up.path
+            # On Windows, this produces ``path == "/C:/path/to/file.ptx"``. Remove the slash.
+            if is_win:
+                path = path[1:]
+            # Use ``resolve()`` to standardize capitalization on Windows.
+            stdpath = pathlib.Path(path).resolve()
+            # Make this path relative to the project directory, to avoid writing potentially confidential information (username / local filesystem paths) to the mapping file, which might be published to the web.
+            relpath = stdpath.relative_to(project_path)
+            # Add this XML ID to others for this path.
+            path_to_xml_id[str(relpath)].append(xml_id)
+
+    # Save the result as a JSON file in the ``dest_dir``.
+    (pathlib.Path(dest_dir) / "mapping.json").write_text(json.dumps(path_to_xml_id))
```

Comment:
Should this be hidden (`.mapping.json`)?

Reply:
Sure, I'll be happy to change the name.